porch/pkg/meta/store.go

```diff
@@ -0,0 +1,229 @@
+	Delete(ctx context.Context, namespacedName types.NamespacedName) (PackageRevisionMeta, error)
+}
+
+// PackageRevisionMeta contains metadata about a specific PackageRevision. The
+// PackageRevision is identified by the name and namespace.
+type PackageRevisionMeta struct {
+	Name        string
+	Namespace   string
+	Labels      map[string]string
+	Annotations map[string]string
+}
+
+var _ MetadataStore = &crdMetadataStore{}
+
+func NewCrdMetadataStore(coreClient client.Client) *crdMetadataStore {
+	return &crdMetadataStore{
+		coreClient: coreClient,
+	}
+}
+
+// crdMetadataStore is an implementation of the MetadataStore interface that
+// stores metadata in a CRD.
+type crdMetadataStore struct {
+	coreClient client.Client
+}
+
+func (c *crdMetadataStore) Get(ctx context.Context, namespacedName types.NamespacedName) (PackageRevisionMeta, error) {
+	ctx, span := tracer.Start(ctx, "crdMetadataStore::Get", trace.WithAttributes())
+	defer span.End()
+
+	var internalPkgRev internalapi.PkgRevMeta
+	err := c.coreClient.Get(ctx, namespacedName, &internalPkgRev)
+	if err != nil {
+		return PackageRevisionMeta{}, err
+	}
+
+	labels := internalPkgRev.Labels
+	delete(labels, PkgRevisionRepoLabel)
+
+	return PackageRevisionMeta{
+		Name:        internalPkgRev.Name,
+		Namespace:   internalPkgRev.Namespace,
+		Labels:      labels,
+		Annotations: internalPkgRev.Annotations,
+	}, nil
+}
+
+func (c *crdMetadataStore) List(ctx context.Context, repo *configapi.Repository) ([]PackageRevisionMeta, error) {
+	ctx, span := tracer.Start(ctx, "crdMetadataStore::List", trace.WithAttributes())
+	defer span.End()
+
+	var internalPkgRevList internalapi.PkgRevMetaList
+	err := c.coreClient.List(ctx, &internalPkgRevList, client.InNamespace(repo.Namespace), client.MatchingLabels(map[string]string{PkgRevisionRepoLabel: repo.Name}))
+	if err != nil {
+		return nil, err
+	}
+	var pkgRevMetas []PackageRevisionMeta
+	var names []string
+	for _, ipr := range internalPkgRevList.Items {
+		labels := ipr.Labels
+		delete(labels, PkgRevisionRepoLabel)
+		pkgRevMetas = append(pkgRevMetas, PackageRevisionMeta{
+			Name:        ipr.Name,
+			Namespace:   ipr.Namespace,
+			Labels:      labels,
+			Annotations: ipr.Annotations,
+		})
+		names = append(names, ipr.Name)
+	}
+	return pkgRevMetas, nil
+}
+
+func (c *crdMetadataStore) Create(ctx context.Context, pkgRevMeta PackageRevisionMeta, repo *configapi.Repository) (PackageRevisionMeta, error) {
+	ctx, span := tracer.Start(ctx, "crdMetadataStore::Create", trace.WithAttributes())
+	defer span.End()
+
+	labels := pkgRevMeta.Labels
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	labels[PkgRevisionRepoLabel] = repo.Name
+	internalPkgRev := internalapi.PkgRevMeta{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        pkgRevMeta.Name,
+			Namespace:   pkgRevMeta.Namespace,
+			Labels:      labels,
+			Annotations: pkgRevMeta.Annotations,
+			// We probably should make these owner refs point to the PackageRevision CRs instead.
```

Comment:
Question, couldn't we have owner refs to both the PackageRevision and repo here?